src/librustc_typeck/check/op.rs

```diff
@@ -186,6 +186,34 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             "binary operation `{}` cannot be applied to type `{}`",
                             op.node.as_str(),
                             lhs_ty);
+
+                        match lhs_ty.sty {
+                            TypeVariants::TyRef(_, ref ty_mut) => {
+                                match ty_mut.ty.sty {
+                                    TypeVariants::TyRef(_, ref ty_mut) => {
+                                        match self.lookup_op_method(expr, ty_mut.ty, vec![rhs_ty_var],
```

Comment:
And at that point the whole thing can probably be generalized to even match the case of a single reference too much (e.g., `&String + &str`).